src/cinder/block_store.go

```diff
@@ -0,0 +1,259 @@
+	}
+	if b.snapshots == nil {
+		b.snapshots = make(map[string]snapshots.Snapshot)
+	}
+
+	// Authenticate to Openstack
+	err := utils.Authenticate(&b.provider, b.log)
+	if err != nil {
+		return fmt.Errorf("failed to authenticate against openstack: %v", err)
+	}
+
+	if b.client == nil {
+		region := utils.GetEnv("OS_REGION_NAME", "")
+		b.client, err = openstack.NewBlockStorageV3(b.provider, gophercloud.EndpointOpts{
+			Region: region,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to create cinder storage object: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// CreateVolumeFromSnapshot creates a new volume in the specified
+// availability zone, initialized from the provided snapshot,
+// and with the specified type.
+// IOPS is ignored as it is not used in Cinder.
+func (b *BlockStore) CreateVolumeFromSnapshot(snapshotName, volumeType, volumeAZ string, iops *int64) (string, error) {
+	b.log.Infof("CreateVolumeFromSnapshot called", snapshotName, volumeType, volumeAZ)
+	readyTimeout := 300
+	snapshotID := b.snapshots[snapshotName].ID
+	volumeName := snapshotID + ".backup." + strconv.FormatUint(rand.Uint64(), 10)
+
+	// Make sure snapshot is in ready state
+	// Possible values for snapshot state:
+	//   https://github.com/openstack/cinder/blob/master/api-ref/source/v3/volumes-v3-snapshots.inc#volume-snapshots-snapshots
+	b.log.Infof("Waiting for snapshot to be in 'available' state", snapshotName, snapshotID, readyTimeout)
+
+	err := snapshots.WaitForStatus(b.client, snapshotID, "available", readyTimeout)
+	if err != nil {
+		b.log.Errorf("snapshot didn't get into 'available' state within the time limit", snapshotName, snapshotID, readyTimeout)
+		return "", err
+	}
+	b.log.Infof("Snapshot is in 'available' state", snapshotName, snapshotID)
+
+	// Create Cinder Volume from snapshot (backup)
+	b.log.Infof("Starting to create volume from snapshot")
+	opts := volumes.CreateOpts{
+		Description:      "Velero backup from snapshot",
+		Name:             volumeName,
+		VolumeType:       volumeType,
+		AvailabilityZone: volumeAZ,
+		SnapshotID:       snapshotID,
+	}
+
+	var cinderVolume *volumes.Volume
+	cinderVolume, err = volumes.Create(b.client, opts).Extract()
+	if err != nil {
+		b.log.Errorf("failed to create volume from snapshot", snapshotName, snapshotID)
+		return "", errors.WithStack(err)
+	}
+	b.log.Infof("Backup volume was created", cinderVolume.ID)
+
+	// Remember the volume
+	b.volumes[cinderVolume.ID] = volumes.Volume{
+		Description:      "Velero backup from snapshot",
+		VolumeType:       volumeType,
+		AvailabilityZone: volumeAZ,
+		SnapshotID:       snapshotID,
+		ID:               cinderVolume.ID,
+		Name:             volumeName,
+	}
+
+	return cinderVolume.ID, nil
+}
+
+// GetVolumeInfo returns the type the specified volume in the given availability zone.
+// IOPS is not used as it is not supported by Cinder.
+func (b *BlockStore) GetVolumeInfo(volumeID, volumeAZ string) (string, *int64, error) {
```

Comment:
volumeAZ is not used in the function

Reply:
also why are you returning *int64 as  second argument? Every return has it set as nil.

Reply:
Added comment that it is not supported by Cinder.